app/upgrade-poolChange.go

```diff
@@ -0,0 +1,96 @@
+package app
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/Sifchain/sifnode/x/clp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/upgrade"
+	"io/ioutil"
+)
+
+func GetPoolChangeFunc(app *SifchainApp) func(ctx sdk.Context, plan upgrade.Plan) {
+	return func(ctx sdk.Context, plan upgrade.Plan) {
+		ctx.Logger().Info("Starting to execute upgrade plan for pool re-balance")
+
+		ExportAppState("changePoolFormula", app, ctx)
+
+		allPools := app.clpKeeper.GetPools(ctx)
+		lps := clp.LiquidityProviders{}
+		poolList := clp.Pools{}
+		hasError := false
+		for _, pool := range allPools {
+			lpList := app.clpKeeper.GetLiquidityProvidersForAsset(ctx, pool.ExternalAsset)
+			temp := sdk.ZeroUint()
+			tempExternal := sdk.ZeroUint()
+			tempNative := sdk.ZeroUint()
+			for _, lp := range lpList {
+				withdrawNativeAssetAmount, withdrawExternalAssetAmount, _, _ := clp.CalculateWithdrawal(pool.PoolUnits, pool.NativeAssetBalance.String(),
+					pool.ExternalAssetBalance.String(), lp.LiquidityProviderUnits.String(), sdk.NewUint(clp.MaxWbasis).String(), sdk.NewInt(0))
+				newLpUnits, lpUnits, err := clp.CalculatePoolUnits(pool.ExternalAsset.Symbol, temp, tempNative, tempExternal,
+					withdrawNativeAssetAmount, withdrawExternalAssetAmount)
+				if err != nil {
+					hasError = true
+					ctx.Logger().Error(fmt.Sprintf("failed to calculate pool units for | Pool : %s | LP %s ", pool.String(), lp.String()))
+					break
+				}
+				lp.LiquidityProviderUnits = lpUnits
+				if !lp.Validate() {
+					hasError = true
+					ctx.Logger().Error(fmt.Sprintf("Invalid | LP %s ", lp.String()))
+					break
+				}
+				lps = append(lps, lp)
+				tempExternal = tempExternal.Add(withdrawExternalAssetAmount)
+				tempNative = tempNative.Add(withdrawNativeAssetAmount)
+				temp = newLpUnits
+			}
+			pool.PoolUnits = temp
+			if !app.clpKeeper.ValidatePool(pool) {
+				hasError = true
+				ctx.Logger().Error(fmt.Sprintf("Invalid | Pool %s ", pool.String()))
+				break
+			}
+			poolList = append(poolList, pool)
+		}
+		// If we have error dont set state
+		if hasError {
+			ctx.Logger().Error("Failed to execute upgrade plan for pool re-balance")
```

Comment:
Just return here and then you can avoid the `!hasError` check below.

Reply:
Yeah , this can be added in , but this is an older upgrade which already went through .